### PR TITLE
Pre build rollup dependencies independent of build pipeline

### PR DESCRIPTION
**Pull request by arthirm**
Enabling the rollup to be prebuilt when the addons are published to artifactory. This prevents the artifacts to be rebuilt every time app is started. 
Later when the addon makes a call to ember-rollup, a check is done to see if the prebuilt directory is already existing, if so returns the prebuilt directory instead of doing rollup.

### CODE REVIEW

**Comment by arthirm** [comment]
@stefanpenner @rwjblue @asakusuma

File: src/index.js [hunk 1]
```diff
@@ -1,35 +1,24 @@
...
+}
 
-  if (namespacedDependencies.length > 0) {
+module.exports = function(modules, indexObj) {
+  const classifyDependencies = require('./rollup-tree').classifyDependencies;
+  var dependencies = classifyDependencies(modules);
+  if (dependencies.namespacedDependencies.length > 0) {
     function treeForAddon(root) {
-      return rollupIntoTree.call(this, root, namespacedDependencies, this._super.treeForAddon);
+      if(indexObj.PREBUILT_PATH) {
```
Comment (stefanpenner) [inline src/index.js]: rather then duplicating this, lets make `addon.isPrebuilt() { .. }` method, and invoke that method here, and in `treeForVendor`

File: src/prebuild.js [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+const path = require('path');
+const builder = require('broccoli-builder');
+const rollUpTree = require('./rollup-tree');
+
+module.exports.preBuild = function(modules, preBuildPath, addonPath) {
+    let treeForAddonPromise, treeForVendorPromise;
+    fs.removeSync(preBuildPath);
+
+    const dependencies = rollUpTree.classifyDependencies(modules);
+    if(dependencies.namespacedDependencies.length > 0) {
```
Comment (stefanpenner) [inline src/prebuild.js]: as node is only single threaded, I think we should build these two sequentially, not in parallel and prebuild should return a single promise for when prebuild is done.

File: src/prebuild.js [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+        let addonTree = rollUpTree.rollUp(dependencies.namespacedDependencies, path.join( addonPath, 'node_modules'));
+        results.push(this.build(addonTree, path.join(preBuildPath, "addon")));
+    }
+    if(dependencies.nonNamespacedDependencies.length > 0) {
+        let vendorTree = rollUpTree.rollUp(dependencies.nonNamespacedDependencies, path.join( addonPath, 'node_modules'));
+        results.push(this.build(vendorTree, path.join(preBuildPath, "vendor")));
+     }
+    return Promise.all(results);
```
Comment (stefanpenner) [inline src/prebuild.js]: these need to run in sequence, not in parallel. This means, chaining such as:

```
buildStepOne().then(() => {
   return buildStepTwo();
});
```

**Comment by Turbo87** [comment]
@arthirm what is `artifactory`? is this just relevant internally at LinkedIn or will this also bring any advantages to other users?

**Comment by stefanpenner** [comment]
> @arthirm what is artifactory? is this just relevant internally at LinkedIn or will this also bring any advantages to other users?

TL;DR NPM === artifactory. This work would be applied during pre-publish step, resulting in pre-built artifacts living within the published tarball
